godot-core/src/obj/raw.rs

```diff
@@ -0,0 +1,624 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::ptr;
+
+use godot_ffi as sys;
+use sys::{interface_fn, GodotFfi, GodotNullableFfi, PtrcallType};
+
+use crate::builtin::meta::{
+    ClassName, FromGodot, GodotCompatible, GodotFfiVariant, GodotType, ToGodot,
+};
+use crate::builtin::{Variant, VariantConversionError};
+use crate::obj::dom::Domain as _;
+use crate::obj::mem::Memory as _;
+use crate::obj::{dom, mem, GodotClass};
+use crate::obj::{GdMut, GdRef, InstanceId};
+use crate::storage::InstanceStorage;
+use crate::{engine, out};
+
+/// Low-level bindings for object pointers in Godot.
+///
+/// This should not be used directly, you should either use [`Gd<T>`](super::Gd) or [`Option<Gd<T>>`]
+/// depending on whether you need a nullable object pointer or not.
+#[repr(C)]
+pub struct RawGd<T: GodotClass> {
+    pub(super) obj: *mut T,
```

Comment:
We currently use `OpaqueObject` here. Does `*mut T` make things easier?
Another option would be `sys::GDExtensionObjectPtr`, but I think this is OK.

Could you add a new `static_assert_eq_size!` next to the existing one, comparing `*mut ()` and `sys::GDExtensionObjectPtr`?

Reply:
Transmuting between pointers and integers can be complicated, <https://github.com/rust-lang/unsafe-code-guidelines/issues/286>. and so i figured it'd be best to avoid doing it if we dont need to.

Reply:
> Also i can add that assert but I'd be very concerned if we're onna platform where pointers can have different sizes.

Fair point 🙂 and it's pretty certain that `GDExtensionObjectPtr` is always going to be a pointer, so there's no real need to.

Reply:
Btw, one nice thing about using `*mut T` here is that we don't need `PhantomData`.